graph/patterns/upsert.md

```diff
@@ -0,0 +1,209 @@
+Customers want to adopt IaC practices for many of the resources managed through Microsoft Graph.
+
+Most resources' creation operations in Microsoft Graph are not idempotent in nature.
+As a consequence, API consumers that want to offer IaC solutions, must create compensation layers that can mimic idempotent behavior. 
+For example, when creating a resource, the compensation layer must check whether the resource first exists, before trying to create or update the resource.
+
+Additionally, IaC code scripts or templates usually employ client-provided names (or keys) to track resources in a predictable manner, whereas [Microsoft Graph guidelines](../GuidelinesGraph.md#behavior-modeling) suggests use of `POST` to create new entities with service-generated keys.
+
+## Solution
+
+The solution is to use an `Upsert` pattern, to solve for the non-idempotent creation and client-provided naming problems.
+
+* For IaC scenarios, resources must use `Upsert` semantics with an [alternate key](./alternate-key.md):
+  * Use `PATCH` with a client-provided alternate key.
+  * For a non-existent resource (specified by the alternate key) the service must handle this as a "create". As part of creation, the service must still generate the primary key value.
+  * For an existing resource (specified by the alternate key) the service must handle this as an "update.
+  * Any new alternate key, used for IaC scenarios, should be called `uniqueName`, if there isn't already an existing property that could be used as an alternate key.
+* NOTE: the service must also support `GET` using the alternate key pattern.
+* For consistent CRUD Microsoft Graph behaviors, all resources, **including** resources used in IaC scenarios, should use `POST` and a service-generated primary key, per existing guidelines, and support `GET`, `PATCH` and `DELETE` using the primary key.
+* If a service does not support `Upsert`, then a `PATCH` call against a non-existent resource must result in an HTTP "409 conflict" error.
+
+This solution allows for existing resources that follow Microsoft Graph conventions for CRUD operations to add `Upsert` without impacting existing apps or functionality.
+
+Ideally, all new entity types should support an `Upsert` mechanism, especially where they support control-plane APIs, or are used in admin style or IaC scenarios.
+
+## When to use this pattern
+
+This pattern should be adopted for resources that are managed through infrastructure as code or desired state configuration.
+
+## Issues and considerations
+
+* The addition of this new pattern (with alternate key) does not represent a breaking change.
+However, some API producers may have concerns about accidental usages of this new pattern unwittingly creating many new resources when the intent was an update.
+As a result, API producers can use the `Prefer: idempotent` to require clients to opt-in to the Upsert behavior.
```

Comment:
>Prefer: idempotent` [](http://example.com/codeflow?start=40&length=19)

Not a fan of this new preference.
1) Is it really a preference? Can services ignore?
2) Is this something we would register w/IANA?
3) This is the same as providing If-Match="*" to force update-only, except has the opposite default, right?

I would rather existing workloads that are worried about backward compatibility not support upsert for existing client-specified keys; I think there are few instances of this.

Reply:
@mikepizzo we had this conversation multiple times.  Not sure where to take this.  It never got resolved with a suitable option, and Graph API council were comfortable with a Prefer header here.

We implemented the prefer header, because If-Match="*" didn't suit the purpose, and the Directory team insisted on some header to indicate this behaviour, over their previous default.

I think we may actually remove this prefer header before we go GA, as services are more comfortable with the approach. So I don't want to register this with IANA.

At the moment, the only client likely to use this is the Graph Bicep Extension.  Other client will likely be a controlled set.  We would recommend they all use the Prefer header.

Reply:
As per discussion in API Council; I don't think this should be called "Idempotent", but I am okay introducing a new preference for:
-existing services with client defined keys
-want to start supporting upsert
-are worried about

Reply:
Updating...